python/test/unit/operators/test_matmul.py

```diff
@@ -137,22 +146,34 @@ def maybe_upcast(x, dtype, is_float8):
             return f8_to_f16(x, dtype)
         return x
 
-    def init_input(m, n, dtype):
+    string_to_torch_type = {"float16": torch.float16, "bfloat16": torch.bfloat16, "float32": torch.float32}
+
+    def init_input(m, n, dtype, only_small_integers=False):
         if 'float8' in dtype:
             ewidth = {'float8e4b15': 4, 'float8e4nv': 4, 'float8e5': 5}[dtype]
             sign = torch.randint(2, size=(m, n), device="cuda", dtype=torch.int8) * 128
             val = torch.randint(2**3 - 1, size=(m, n), device="cuda", dtype=torch.int8) << 7 - ewidth
             return sign | val
         if dtype == "int8":
+            if only_small_integers:
+                return torch.randint(-2, 2, (m, n), device="cuda", dtype=torch.int8)
             return torch.randint(-128, 127, (m, n), device="cuda", dtype=torch.int8)
-        dtype = {"float16": torch.float16, "bfloat16": torch.bfloat16, "float32": torch.float32}[dtype]
-        exponents = torch.randint(-10, 0, size=(m, n))
-        ret = (2. ** exponents).to(dtype).to("cuda")
+        if only_small_integers:
+            exponents = torch.randint(0, 3, size=(m, n))
+        else:
+            exponents = torch.randint(-10, 0, size=(m, n))
+        ret = (2. ** exponents).to(string_to_torch_type[dtype]).to("cuda")
         return ret
 
     # allocate/transpose inputs
-    a = init_input(M, K, ADTYPE)
-    b = init_input(K, N, BDTYPE)
+    use_small_integers = False
+    # We can't force torch to use float16 for internal computations (it will use float32), and
```

Comment:
I think that currently it IS the issue. Epsilon from float16 is around ~1e-3, way larger than the maximum expected error of 1e-5 so there is no chance that they will match using non-integers. So, it's never going to work. 

But I agree about that even using float16 both sides we could have potentially the issue because floating point arithmetic and associativity. I will rephrase the comment.